Small-to-Large Merging Solution and Editorial for IOI Race (#2759)
cpinitiative/usaco-guide: solutions/ioi-11-race.mdx

To accomplish this, we can use small to large merging. For each $i$, we maintain a map of key to value pair of $\text {sum}(root, j) : \text {dist}(root, j)$ for  distinct $\text {sum}(root, j)$ for all $j$ in the subtree of $i$. If there are duplicate $\text {sum}(root, j)$ we take the minimum $\text {dist}(root, j)$.

As we combine the maps of two nodes, we iterate through each key-pair $(sum_s, dist_s)$ of the smaller size map, search for a key-pair $sum_l = s - sum_s$ in the larger map, and if this exists we obtain $dist_s + dist_l$ from $sum_l$. Our answer would be minimum $dist_s + dist_l$ over all mergings. During this process we also merge all key-value pairs of the smaller map.

## Implementation

**Time Complexity:** $\mathcal{O}(N\log^2N)$

The log factors come from small-to-large merging and the map operations.

```cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;
typedef pair<ll, ll> pii;

#define pb push_back

#define f first
#define s second

const int maxn = 2e5 + 1;
vector<pii> adj[maxn];
map<ll, ll> info[maxn];
ll dist[maxn];
ll sum[maxn];  // define these relative to root

> int N, K, u, v;

Reviewer: wait i don't think `u` and `v` even have to be global?

Author: true, changed

---
id: ioi-11-race
source: IOI 2011
title: Race
author: Timothy Gao
---

[Official Analysis (Centroid Decomposition)](https://ioinformatics.org/page/ioi-2011/37)

## Alternative Solution: Small to Large Merging

We are tasked with finding a path with the minimum number of traversed edges such that the path sum is $s$.
First, we root the tree arbitrarily.

> Let $lca(u, v) = w$.

Reviewer 2: `\texttt lca(u, v)`

> We can break $path(u, v)$ into $path(u, w)$ and $path(v, w)$.

Reviewer 2: same here, use `\texttt`

Reviewer 2: not resolved

Author: added a bunch of \text {}

We perform small to large merging to find an $lca$ $i$ such that $sum(u, root) + sum(v, root) - 2 \cdot sum(i, root) = s$, and $dist(u, root) + dist(v, root) - 2 \cdot dist(i, root) = s$ is minimized. 

To accomplish this, we can use small to large merging. For each $i$, we maintain a map of key to value pair of $sum(root, j) : dist(root, j)$ for  distinct $sum(root, j)$ for all $j$ in the subtree of $i$. If there are duplicate $sum(root, j)$ we take the minimum $dist(root, j)$.

As we combine the maps of two nodes, we iterate through each key-pair $(sum_s, dist_s)$ of the smaller size map, search for a key-pair $sum_l = s - sum_s$ in the larger map, and if this exists we obtain $dist_s + dist_l$ from $sum_l$. Our answer would be minimum $dist_s + dist_l$ over all mergings. During this process we also merge all key-value pairs of the smaller map.

## Implementation

**Time Complexity:** $\mathcal{O}(N\log^2N)$

The log factors come from small-to-large merging and the map operations.

```cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;
typedef pair<ll, ll> pii;

#define pb push_back

#define f first
#define s second

> const int maxn = 2e5 + 20;

Reviewer: 1. MAX_N instead of maxn
2. would just `2e5` work?

vector<pii> adj[maxn];
map<ll, ll> info[maxn];
ll dist[maxn];

> ll sum[maxn]; // define these relative to root

Reviewer: two spaces after code before starting an inline comment
```suggestion
ll sum[maxn];  // define these relative to root
```

int N;
int K;
int u;
int v;
ll ret;

void dfs1(int u, int p, ll c, int h)
{

>     info[u][c] = h;

Reviewer: how could i forget
tabs

> int v;

Reviewer: bruh no

    sum[u] = c;
    dist[u] = h;
    for (auto n : adj[u])
    {
        if (p == n.f)
        {
            continue;
        }
        dfs1(n.f, u, c + n.s, h + 1);
    }
}

> void dfs(int u, int p)

Reviewer: preferably a better name, what does this `dfs` do?

---
id: ioi-11-race
source: IOI 2011
title: Race
author: Timothy Gao
---

> [Official Analysis with Centroid Decomposition Downloaded Here](https://ioinformatics.org/page/ioi-2011/37)

Reviewer: ```suggestion
[Official Analysis (Centroid Decomposition)](https://ioinformatics.org/page/ioi-2011/37)
```

We perform small to large merging to find an $lca$ $i$ such that $sum(u, root) + sum(v, root) - 2 \cdot sum(i, root) = s$, and $dist(u, root) + dist(v, root) - 2 \cdot dist(i, root) = s$ is minimized. 

To accomplish this, we can use small to large merging. For each $i$, we maintain a map of key to value pair of $sum(root, j) : dist(root, j)$ for  distinct $sum(root, j)$ for all $j$ in the subtree of $i$. If there are duplicate $sum(root, j)$ we take the minimum $dist(root, j)$.

As we combine the maps of two nodes, we iterate through each key-pair $(sum_s, dist_s)$ of the smaller size map, find if there exists a key-pair $sum_l = s - sum_s$ in the larger map, and obtain $dist_s + dist_l$. Our answer would be minimum $dist_s + dist_l$ over all mergings. During this process we also merge all key-value pairs of the smaller map.

## Implementation

**Time Complexity:** $\mathcal{O}(N\log^2N)$

> One log factor comes from small-to-large merging, and the other comes from the map operations.

Reviewer 3: ```suggestion
The log factors come from small-to-large merging and the map operations.
```

```cpp
#include <bits/stdc++.h>
 
using namespace std;
 
typedef long long ll;
typedef pair<ll, ll> pii;
 
#define pb push_back
 
#define f first
#define s second
 
const int maxn = 2e5+20;
vector<pii> adj[maxn];
map<ll, ll> info[maxn]; 
ll dist[maxn], sum[maxn]; //define these relative to root

> int N, K, u, v; ll ret;

Reviewer: declarations on separate lines

 
void dfs1(int u, int p, ll c, int h){
	info[u][c] = h; 

> 	sum[u] = c; dist[u] = h;

Reviewer: statements on separate lines

	for(auto n : adj[u]){

> 		if(p == n.f) continue;

Reviewer: braces (other places too)

		dfs1(n.f, u, c + n.s, h + 1);
	}
}
 
void dfs(int u, int p){

> 	ll target = K + 2*sum[u];

Reviewer: spaces around operators

> 	for(auto n : adj[u]){

Reviewer: spaces for control flow statements
```suggestion
	for (auto n : adj[u]) {
```
and others

		if(n.f == p) continue;
		dfs(n.f, u);
		if(info[n.f].size() > info[u].size()){
			swap(info[n.f], info[u]);
		}
		for(auto i : info[n.f]){
			if(info[u].find(target - i.f) != info[u].end()){
				ret = min(ret, info[u][target - i.f] + i.s - 2*dist[u]);
			}
		}
		for(auto i : info[n.f]){
			if(info[u].find(i.f) == info[u].end()){
				info[u].insert(i);
			}else{
				info[u][i.f] = min(info[u][i.f], i.s);
			}
		}
	} 
}

int best_path(int n, int k, int edges[][2], int weights[]){

> 	if(k == 1) return 0;

Reviewer 3: add braces
